proposals/devcontainer-features.md

```diff
@@ -0,0 +1,143 @@
+# Dev container features reference
+
+Dev container features provide a smooth path for customizing your container definitions. Features are self contained units of code meant to facilitate creation of custom images for application or development containers.
+
+From a practical point of view, features are folders that contain units of code with different entrypoints for different lifecycle events.
+
+Features can be defined by a `feature.json` file in the root folder of the feature. The file is optional for backwards compatibility but it is required for any new features being authored.
```

Comment:
Let's use a distinctive name to make it easy to associate a schema file (and possibly other editing support) to it. `devcontainer-feature.json` would work well. Similarly, let's use `devcontainer-feature-collection.json` or shorter `devcontainer-features.json`.

Reply:
I liked the shorthand of just `feature.json` but if we agree on that change, I can make it.

Reply:
Made the changes. cc @joshspicer, @samruddhikhandale